### PR TITLE
!!! TASK: Streamline `ContentGraph::findRootNodeAggregateByType`

### PR DESCRIPTION
**Pull request by mhsdesign**
to be nullable like findRootNodeByType in the subgraph

**Upgrade instructions**

**Review instructions**

I wondered why is `findRootNodeByType` in the subgraph nullable while `findRootNodeAggregateByType` in the content graph is not nullable and throws an `RootNodeAggregateDoesNotExist` exception? That seems the only place where we throw deliberately exceptions in the read side if a user of the cr put wrong stuff in the query.

1. `findRootNodeAggregateByType` was nullable introduced via https://github.com/neos/neos-development-collection/commit/1f4b464ec6e5c515ae94c240205fbcced077a907#diff-a623327bb4acbff32c1cbab65ef64b94e682baddd7505e56ca5f08b08e80ecc2

2. nullability was removed via https://github.com/neos/neos-development-collection/commit/8ab8a1b4a794cab412a9b631c1ea3920db95527f#diff-a623327bb4acbff32c1cbab65ef64b94e682baddd7505e56ca5f08b08e80ecc2

last refactoring https://github.com/neos/neos-development-collection/pull/4339 and https://github.com/neos/neos-development-collection/pull/4129

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by kitsunet** [comment]
Mmm, I guess there are arguments in both directions, I generally like it as is, with the Exception, but it makes sense to have things similar between the subgraph and the content graph.

**Comment by mhsdesign** [comment]
jip my argument would be that there is no `find` method that throws if you try to find something that doesnt exist (because of bad user input).

The only thing that throws is `getContentGraph` but we discussed that we liked it there.

Handling nullability yourself is also way easier than knowing which exceptions to catch and which not... phpstan will help you. Im really glad that `findParent` etc are now nullable (unlike once proposed in the traversableNodeInterface) and this should probably be aligned to behave the same.
The question is, why was it nullable once and that has been removed??!

**Comment by kitsunet** [comment]
Good question yes, I guess in some cases were the error case is fine you can skip an additional null check, but fine for me to merge it, we need ot adapt the use cases as well though!

File: Neos.Neos/Classes/Domain/Service/SiteServiceInternals.php [hunk 1]
```diff
@@ -60,6 +60,10 @@ public function removeSiteNode(SiteNodeName $siteNodeName): void
             $sitesNodeAggregate = $contentGraph->findRootNodeAggregateByType(
                 NodeTypeNameFactory::forSites()
             );
+            if (!$sitesNodeAggregate) {
+                // nothing to prune, we could probably also return here directly?
```
Comment (kitsunet) [inline Neos.Neos/Classes/Domain/Service/SiteServiceInternals.php:64]: That would be the current behavior with the exception I guess? Or rather we might actually throw here to tell people that this site does not exist?

File: Neos.Neos/Classes/Controller/Module/Administration/SitesController.php [hunk 2]
```diff
@@ -184,11 +184,10 @@ public function updateSiteAction(Site $site, $newSiteNodeName)
...
 
-            try {
-                $sitesNode = $contentRepository->getContentGraph($liveWorkspace->workspaceName)->findRootNodeAggregateByType(
-                    NodeTypeNameFactory::forSites()
-                );
-            } catch (\Exception $exception) {
+            $sitesNode = $contentRepository->getContentGraph($liveWorkspace->workspaceName)->findRootNodeAggregateByType(
+                NodeTypeNameFactory::forSites()
+            );
+            if (!$sitesNode) {
```
Comment (kitsunet) [inline Neos.Neos/Classes/Controller/Module/Administration/SitesController.php]: === null

File: Neos.ContentRepository.Core/Classes/Service/ContentRepositoryBootstrapper.php [hunk 3]
```diff
@@ -62,27 +62,25 @@ public function getOrCreateLiveWorkspace(): Workspace
...
-            $this->contentRepository->handle(CreateRootNodeAggregateWithNode::create(
-                $workspace->workspaceName,
-                $rootNodeAggregateId,
-                $rootNodeTypeName,
-            ));
-            return $rootNodeAggregateId;
+        $rootNodeAggregate = $this->contentRepository->getContentGraph($workspace->workspaceName)->findRootNodeAggregateByType(
+            $rootNodeTypeName
+        );
+        if ($rootNodeAggregate) {
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/Service/ContentRepositoryBootstrapper.php]: !== null

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/Service/ContentRepositoryBootstrapper.php]:
```suggestion
        if ($rootNodeAggregate !== null) {
```

File: Neos.Neos/Classes/Controller/Module/Administration/SitesController.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/Controller/Module/Administration/SitesController.php]:
```suggestion
            if ($sitesNode === null) {
```

File: Neos.Neos/Classes/Domain/Service/SiteServiceInternals.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/Domain/Service/SiteServiceInternals.php:64]: yes we throw currently. But as we want to remove a child node either way it seems fine to say, well we cant remove a thing here.

Comment (bwaidelich) [reply Neos.Neos/Classes/Domain/Service/SiteServiceInternals.php:64]: I think, skipping is the better behavior.
But maybe we could throw at the end of this method if no node was actually removed?

File: Neos.TimeableNodeVisibility/Classes/Service/TimeableNodeVisibilityService.php [hunk 4]
```diff
@@ -101,7 +98,7 @@ private function getNodesWithExceededDates(ContentRepository $contentRepository,
             $sitesNodeTypeName = NodeTypeName::fromString('Neos.Neos:Sites');
             $rootNode = $subgraph->findRootNodeByType($sitesNodeTypeName);
             if ($rootNode === null) {
-                throw RootNodeAggregateDoesNotExist::butWasExpectedTo($sitesNodeTypeName);
+                throw new \RuntimeException(sprintf('No sites root node found in content repository "%s"', $contentRepository->id->value), 1719047148);
```
Comment (bwaidelich) [inline Neos.TimeableNodeVisibility/Classes/Service/TimeableNodeVisibilityService.php:101]: Just wondering: Why don't we keep the `RootNodeAggregateDoesNotExist` exception and throw it here? (not sure if it's needed but I'm curious about your reasons to replace it)

**Review by bwaidelich — APPROVED** [review]
> The question is, why was it nullable once and that has been removed??!

It was never nullable, it was introduced with:
https://github.com/neos/neos-development-collection/commit/1f4b464ec6e5c515ae94c240205fbcced077a907#diff-a623327bb4acbff32c1cbab65ef64b94e682baddd7505e56ca5f08b08e80ecc2

But it replaced the previous `findRootNodeByType` that _was_ nullable.

> Mmm, I guess there are arguments in both directions

I'm also torn sometimes between exception and nullable, since the latter is slightly less explicit.
But for these cases it totally makes sense IMO (as you can tell from the simplified resulting code) and also we had that soft guideline that `find*` returns nullable types and `get*` throws.

Just some minor remarks that might be ignored

File: Neos.TimeableNodeVisibility/Classes/Service/TimeableNodeVisibilityService.php (hunk 4 above)
Comment (mhsdesign) [reply Neos.TimeableNodeVisibility/Classes/Service/TimeableNodeVisibilityService.php:101]: we are in a different package here (the timeable stuff) and its imo not needed here to have a special exception class. Its a not ever gonna happen case and just a save guard.

File: Neos.Neos/Classes/Domain/Service/SiteServiceInternals.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.Neos/Classes/Domain/Service/SiteServiceInternals.php:64]: I dont think so. But with https://github.com/neos/neos-development-collection/issues/4470 we can definitely discuss how the full behaviour should be and adjust edge cases.

**Review by nezaniel — APPROVED** [review]
I prefer having find*ers nullable consistently, so +1 from me.
In this case especially, getOrCreateRootNodeAggregate (see above) gets a lot more comprehensible.
Tests should again be green as well.

**Pull request merged**